sdk/core/core-client/src/storageChallengeAuthentication.ts

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  AuthorizeRequestOnChallengeOptions,
+  PipelineRequest,
+  PipelineResponse,
+} from "@azure/core-rest-pipeline";
+
+import { GetTokenOptions } from "@azure/core-auth";
+
+/**
+ * A set of constants used internally when processing requests.
+ */
+const Constants = {
+  DefaultScope: "/.default",
+  /**
+   * Defines constants for use with HTTP headers.
+   */
+  HeaderConstants: {
+    /**
+     * The Authorization header.
+     */
+    AUTHORIZATION: "authorization",
+  },
+};
+
+/**
+ * Defines a callback to handle auth challenge for Storage APIs
+ **/
+export const storageAuthorizeRequestOnChallenge: (
```

Comment:
Is there a way to name this in a more generic way? Since this is used from more than one kind of client. Something that describes more what this onChallenge callback is doing?

Reply:
Part of the rationale is that there’s nothing in this code that I can see that binds it specifically to storage.

Reply:
This has things that are specific to the [storage spec ](https://docs.microsoft.com/en-us/rest/api/storageservices/authorize-with-azure-active-directory#bearer-challenge), for example value delimiter is spaces vs commas in the general [AAD docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/claims-challenge). Also, storage doesn't have quotes wrapping the string values.

There are other things that differ, for example KV requires an initial request to be sent without a body, which storage doesn't.

Reply:
While it is true that the storage spec follows this approach, are there other Azure service specs that follow that approach?

The characteristics that storage follows, that are embodied with this code, might be more tightly coupled with some underlying security consideration rather than with the storage service.

Reply:
I'm a bit wary of shipping something so tied to storage in `core-client` - perhaps a better solution would be to establish a `storage-common` type package? I believe storage has other areas where they re-use code today too

Reply:
That would be cool, but Tables is not strictly Storage, it also works with Cosmos, they implemented the same Storage Tables API. 

I guess we can still create a storage common package, especially if we publish it. Another option would be to rename it to `crossTenantChallenge` instead of having the storage name in there.

Reply:
I agree with @sadasant that this feels more generic than storage specific (how the scopes and tenant id are stored in challenge seems just implementation detail). Currently we allow providing scopes and tenant id when retrieving aad access tokens. @sadasant do we know whether there are other GetAccessToken options that we don't currently support? We can probably make this generic, and allow more options if there are any in the future